evap/results/exporters.py

```diff
@@ -123,52 +132,58 @@ def filter_evaluations(semesters, evaluation_states, degrees, course_types, cont
                         or contribution_result.contributor is None
                         or contribution_result.contributor == contributor
                     ):
-                        results.setdefault(questionnaire_result.questionnaire.id, []).extend(
-                            questionnaire_result.question_results
-                        )
+                        results.setdefault(questionnaire_result.questionnaire.id, []).extend(question_results)
                         used_questionnaires.add(questionnaire_result.questionnaire)
-            evaluation.course_evaluations_count = evaluation.course.evaluations.count()
-            if evaluation.course_evaluations_count > 1:
+            evaluation_as_any: Any = evaluation
+            evaluation_as_any.course_evaluations_count = evaluation_as_any.course.evaluations.count()
+            if evaluation_as_any.course_evaluations_count > 1:
                 course_results_exist = True
-                weight_sum = sum(evaluation.weight for evaluation in evaluation.course.evaluations.all())
-                evaluation.weight_percentage = int((evaluation.weight / weight_sum) * 100)
-                evaluation.course.avg_grade = distribution_to_grade(
-                    calculate_average_course_distribution(evaluation.course)
+                weight_sum = sum(evaluation.weight for evaluation in evaluation_as_any.course.evaluations.all())
+                evaluation_as_any.weight_percentage = int((evaluation.weight / weight_sum) * 100)
+                evaluation_as_any.course.avg_grade = distribution_to_grade(
+                    calculate_average_course_distribution(evaluation_as_any.course)
                 )
-            evaluations_with_results.append((evaluation, results))
+            evaluations_with_results.append((evaluation_as_any, results))
 
         evaluations_with_results.sort(
             key=lambda cr: (cr[0].course.semester.id, cr[0].course.type.order, cr[0].full_name)
         )
-        used_questionnaires = sorted(used_questionnaires)
+        sorted_questionnaires = sorted(used_questionnaires)
 
-        return evaluations_with_results, used_questionnaires, course_results_exist
+        return evaluations_with_results, sorted_questionnaires, course_results_exist
 
     def write_headings_and_evaluation_info(
-        self, evaluations_with_results, semesters, contributor, degrees, course_types, verbose_heading
-    ):
+        self,
+        evaluations_with_results: list[tuple[Evaluation, OrderedDict[int, list[QuestionResult]]]],
+        semesters: Sequence[Semester] | QuerySet[Semester],
+        contributor: UserProfile | None,
+        degree_ids: Iterable[int],
+        course_type_ids: Iterable[int],
+        verbose_heading: bool,
+    ) -> None:
         export_name = _("Evaluation")
         if contributor:
             export_name += f"\n{contributor.full_name}"
         elif len(semesters) == 1:
             export_name += f"\n{semesters[0].name}"
```

Comment:
If you do
```suggestion
            export_name += f"\n{next(iter(semesters)).name}"
```
you can type `semesters` as `Collection[Semester]`

Reply:
To sum this up: Django is not consistent with `QuerySet`. `QuerySet` implements `contains` but not `__contains__`. This cannot be annotated as a Container because `__contains__` is missing. So using a common base type is not possible. This could be typed as an `Sequence` (with this suggestion), but then typeguard complains about not having the same type, which is obviously true. MyPy does not complain because everything used is implemented and Django Stubs implements `__contains__`. It is possible to be a Container that does not provide `__contains__` this trigger's fallback, so `__iter__` with comparisons is used. Because `__iter__` is defined, we could use `Iterable` but we deemed this as too generic.

Reply:
Please see https://github.com/typeddjango/django-stubs/issues/1924 for the QuerySet Typing Issue

Reply:
The correct type is a class which inherits `Sized` and `Iterable`